snapcraft/internal/meta/command.py

```diff
@@ -0,0 +1,183 @@
+from snapcraft.internal import common
+
+_COMMAND_PATTERN = re.compile("^[A-Za-z0-9. _#:$-][A-Za-z0-9/. _#:$-]*$")
+
+
+def _get_shebang_from_file(file_path: str) -> Optional[str]:
+    """Returns the shebang from file_path."""
+    if not os.path.exists(file_path):
+        raise errors.ShebangNotFoundError()
+
+    with open(file_path, "rb") as exefile:
+        if exefile.read(2) != b"#!":
+            raise errors.ShebangNotFoundError()
+        shebang_line = exefile.readline().strip().decode("utf-8")
+
+    if shebang_line.startswith("/usr/bin/env"):
+        shebang = shebang_line.split()[1]
+    else:
+        shebang = shebang_line
+
+    # if the shebang startswith / we should not extract it.
+    if shebang.startswith("/"):
+        raise errors.ShebangInRoot()
+
+    return shebang
+
+
+def _find_binary(*, binary: str, prime_dir: str) -> str:
+    binary_paths = (
+        os.path.join(p, binary) for p in common.get_bin_paths(root=prime_dir)
+    )
+    found_path = None
+    for binary_path in binary_paths:
+        if _executable_is_valid(binary_path):
+            found_path = binary_path
+            break
+    else:
+        found_path = shutil.which(binary)
+
+    if found_path is None:
+        raise errors.PrimedCommandNotFoundError(binary)
+
+    return found_path
+
+
+def _get_command_path(*, command: str, prime_dir: str) -> str:
+    if command.startswith("/"):
+        command_path = command
+    # Remove a leading $SNAP if found.
+    elif command.startswith("$SNAP/"):
+        command_path = os.path.join(prime_dir, command[len("$SNAP/") :])
+    else:
+        command_path = os.path.join(prime_dir, command)
+
+    return command_path
+
+
+def _massage_command(*, command: str, prime_dir: str) -> str:
+    # If command starts with "/" we have no option but to use a wrapper.
+    if command.startswith("/"):
+        return command
+
+    command_parts = shlex.split(command)
+    command_path = _get_command_path(command=command_parts[0], prime_dir=prime_dir)
+
+    # Extract shebang
+    with contextlib.suppress(errors.ShebangNotFoundError, errors.ShebangInRoot):
+        shebang = _get_shebang_from_file(command_path)
+        # Add the shebang (interpreter) to the front of the command.
+        command_parts.insert(0, shebang)
+        # And make sure the original command is prepended with $SNAP so it is
+        # found.
+        command_parts[1] = os.path.join("$SNAP", command_parts[1])
+        command_path = _get_command_path(command=shebang, prime_dir=prime_dir)
+
+    if not os.path.exists(command_path) and command_parts[0].startswith("$SNAP/"):
+        raise errors.PrimedCommandNotFoundError(command_parts[0])
+    elif not os.path.exists(command_path):
+        command_path = _find_binary(binary=command_parts[0], prime_dir=prime_dir)
+
+    if command_path.startswith(prime_dir):
+        command_parts[0] = os.path.relpath(command_path, prime_dir)
+    else:
+        command_parts[0] = command_path
```

Comment:
I tried to go down this path and it gets weird since to get the interpreter I still need to discover the command and at that point there is no separation of concerns